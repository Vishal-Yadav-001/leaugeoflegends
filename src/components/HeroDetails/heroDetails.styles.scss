@import "../../styles/variables.styles.scss";
.MuiBox-root.box {
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100%;
  background-color: $primary;
  opacity: 0.9;
}

.MuiCard-root.hero_container {
  opacity: 0.9;
  background-color: $black;
  flex-wrap: wrap;
  flex-direction: column;
  position: relative;
}

.MuiCardContent-root img.hero_full_image {
  width: clamp(3rem, 40vw, 12rem);
}

.MuiCardContent-root .MuiBox-root .specs {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  color: $neutral;
 // width: clamp(3rem, 40vw, 12rem);
}
.MuiCardContent-root .MuiBox-root .specs .MuiTypography-root {
 font-size: .8rem;
 font-weight: 400;

}
.logo {
  // width: clamp(1.2rem, 2.5vw, 1.8rem);
}
.MuiIconButton-root.icon {
  color: $accent;
  margin-left: 1rem;
}
.specs_container {
  padding-bottom: 0.5rem;
}
.MuiCardContent-root.description {
  text-align: left;
  font-weight: 300;
  font-size: .9rem;
  line-height: 1rem;

}
.MuiSkeleton-root.heading{
  background-color: $accent;
  height:25rem;
  width:60vh;
}
