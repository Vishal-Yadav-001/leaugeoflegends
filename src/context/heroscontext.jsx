import { createContext, useState } from "react";

const HerosContext = createContext();

const HerosProvider = ({ children }) => {
  const [favoriteHero, setFavoriteHero] = useState([]);
  const [isHeroPresent, setIsHeroPresent] = useState([]);
  /**
   * Add selected hero users watch list
   * @param {*} - value
   */
  const addHero = (value) => {
    setFavoriteHero((prev) => {
      if (prev && prev.includes(value)) {
        return [...prev];
      } else {
        setIsHeroPresent((prev)=> [...prev,value?.id])
        return [...prev, value];
      }
    });
  };
  /**
   * Check if hero is already present in user's watchlist
   * @param {number} id - id of the selected hero
   * @returns {boolean} true if hero already exist else false
   */
  const checkIsHeroPresent = (id) => {
    return isHeroPresent?.includes(id) ? true : false;
  };
  /**
   * Removes selected hero from user's watchlist
   * @param {number} id 
   */
  const handleRemoveRecord = (id)=>{
    setFavoriteHero((prev)=> prev.filter(data=> data.id !== id));
    setIsHeroPresent((prev)=> prev.filter(data=> data !== id));
   }
  return (
    <HerosContext.Provider
      value={{
        favoriteHero,
        setFavoriteHero,
        addHero,
        checkIsHeroPresent,
        isHeroPresent,
        setIsHeroPresent,
        handleRemoveRecord
      }}
    >
      {children}
    </HerosContext.Provider>
  );
};

export { HerosContext, HerosProvider };
