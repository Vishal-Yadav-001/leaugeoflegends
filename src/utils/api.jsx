import axios from "axios";
/**
 * for local dev use https://api.pandascore.co/lol/champions
 */
const API_KEY = "eNraJAvEMOVX0M_d7IfaecVTtJK_QiAXA-_qUL8WJckQX44PZ1U";
const baseUrl = "/api/";

/**
 * @returns {Array.<Object>} list of heros
 */
export const getListOfChampions = async (herosCount) => {
  const pageNumber = 1;
  const pageSize = (herosCount || 50);
  const config = {
    params: {
      "page[number]": pageNumber,
      "page[size]": pageSize,
      token: API_KEY,
    },
  };
  try {
    const response = await axios.get(`${baseUrl}`, config);
    return response?.data;
  } catch (error) {
    throw error;
  }
};
/**
 * @param {string} heroName used to fetch hero details
 */
export const getChampion = async (championName) => {
  const additionalDataURL = "https://lolbe2.azurewebsites.net/api/v1/champstaticdata";
  const config = {
    params: {
      "search[name]": championName,
      token: API_KEY,
    },
  };
  const additionalConfig = {
    params: {
      "champion": championName,
      
    },
  };
  
  try {
    const response = await axios.get(`${baseUrl}`, config);
    const additionalResponse = await axios.get(`${additionalDataURL}`,additionalConfig);
    const introduction = {"description":additionalResponse.data[0]?.introduction,"image":additionalResponse.data[0]?.image,class:additionalResponse.data[0]?.class};
    const finalResponse = [{...response.data[0],...introduction}];
    if(response && additionalResponse){
      return finalResponse;
    }
  } catch (error) {
    throw error;
  }
};
