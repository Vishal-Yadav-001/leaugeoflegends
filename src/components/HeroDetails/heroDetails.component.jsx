import {
  BookmarkAddOutlined,
  BookmarkRemove,
  BookmarkRemoveOutlined,
  Close,
  EmojiPeople,
  Fastfood,
  LocationCity,
} from "@mui/icons-material";
import {
  Backdrop,
  Box,
  Button,
  Card,
  CardActionArea,
  CardActions,
  CardContent,
  CardHeader,
  Divider,
  Fade,
  Grid,
  IconButton,
  Modal,
  Skeleton,
  Stack,
  Tooltip,
  Typography,
  Zoom,
} from "@mui/material";
import { Container } from "@mui/system";
import React, { Fragment, useContext, useEffect, useState } from "react";
import { HerosContext } from "../../context/heroscontext";
import { getChampion } from "../../utils/api";
import Loader from "../Loader/loader.component";
import "./heroDetails.styles.scss";
/**
 * Displays details on individual hero selected by the user
 * @param {string} selectedHero - name of the selected hero
 * @param {boolean} open - to open modal
 *
 */
const HeroDetails = ({ selectedHero, open, setOpen }) => {
  const [heroDetails, setHeroDetails] = useState(null);
  const [loading, setLoading] = useState(false);
  const handleClose = () => setOpen(false);
  const {
    addHero,
    setIsHeroPresent,
    isHeroPresent,
    checkIsHeroPresent,
    handleRemoveRecord,
  } = useContext(HerosContext);

  /**
   * Method for conversion of custom values (numbers) to scale of 0 to 100 for progress bar/circle.
   * @param {number} value - number
   * @returns {number}
   */
  const normalize = (value) => {
    let max = 0;
    let min = 0;
    switch (true) {
      case value < 10:
        max = 10;

        break;
      case value > 10 && value < 101:
        max = 100;
        break;
      case value > 100:
        max = 1000;
        break;
      default:
        break;
    }
    return ((value - min) * 100) / (max - min);
  };
  useEffect(() => {
    if (selectedHero) {
      setLoading(true);
      setHeroDetails(null);
      getChampion(selectedHero).then((data) => {
        setLoading(false);
        setHeroDetails(data);
      });
    }
  }, [selectedHero]);

  // const constructDetails = (object,propertyName)=>{
  //   for(const key in object){
  //     if(key === propertyName){
  //       return key.toUpperCase()
  //     }
  //   }
  // }

  return (
    <Fragment>
      <Modal
        aria-labelledby="transition-modal-title"
        aria-describedby="transition-modal-description"
        open={open}
        onClose={handleClose}
        closeAfterTransition
        BackdropComponent={Backdrop}
        BackdropProps={{
          timeout: 1000,
        }}
      >
        <Fade in={open}>
          <Box className="box">
            {loading ? (
              <Box>
                <Typography
                  variant="h5"
                  textAlign="center"
                  sx={{ color: "#fff" }}
                >
                  Loading...{" "}
                </Typography>
                <Skeleton data-testid="heading" className="heading"></Skeleton>
              </Box>
            ) : (
              heroDetails?.map((data) => {
                return (
                  <div key={data?.id} style={{ width: "60vh" }}>
                    <div>
                      <Card
                        className="hero_container"
                        sx={{
                          display: "flex",
                          flexDirection: "column",
                          maxWidth: "700px",
                        }}
                      >
                        <CardHeader
                          variant="h4"
                          title={
                            <Fragment>
                              <Box sx={{ display: "flex", color: "#FFB100" }}>
                                <Typography variant="h4">
                                  {" "}
                                  {data?.name}
                                </Typography>
                                {checkIsHeroPresent(data?.id) ? (
                                  <Tooltip
                                    TransitionComponent={Zoom}
                                    title="Remove from watchlist"
                                    placement="top"
                                  >
                                    <IconButton
                                      className="icon"
                                      onClick={() =>
                                        handleRemoveRecord(data?.id)
                                      }
                                      data-testid="remove hero"
                                    >
                                      <BookmarkRemove  />
                                    </IconButton>
                                  </Tooltip>
                                ) : (
                                  <Tooltip
                                    TransitionComponent={Zoom}
                                    title="Add to watchlist"
                                    placement="right-start"
                                  >
                                    <IconButton
                                      className="icon"
                                      onClick={() => addHero(data)}
                                      data-testid="add hero"
                                    >
                                      <BookmarkAddOutlined />
                                    </IconButton>
                                  </Tooltip>
                                )}
                                <Box >
                                  {data?.class?.map((heroClass) => {
                                    return (
                                      <Tooltip
                                        TransitionComponent={Zoom}
                                        title={heroClass?.description}
                                        placement="top"
                                      >
                                        <img
                                          style={{ width: "35px",borderRadius:"50%"  }}
                                          alt={heroClass?.description}
                                          src={heroClass?.image}
                                        />
                                      </Tooltip>
                                    );
                                  })}
                                </Box>
                                <Tooltip
                                  TransitionComponent={Zoom}
                                  title="Close"
                                  placement="top"
                                >
                                    <Close sx={{color:"##FFB100",position:"absolute",right:"1rem",top:".5rem"}} onClick={handleClose}/>
                                </Tooltip>
                              </Box>
                              <Divider />
                            </Fragment>
                          }
                        />
                        <Box sx={{ display: "flex" }}>
                          <CardContent>
                            <img
                              className="hero_full_image"
                              alt={data?.name}
                              src={data?.big_image_url}
                              style={{ height: "100%" }}
                            ></img>
                          </CardContent>

                          <CardContent className="description">
                            <Typography variant="span">
                              {data?.description || "No info available"}
                            </Typography>

                            <Divider
                              sx={{
                                paddingTop: ".5rem",
                                paddingBottom: ".5rem",
                              }}
                            />
                            <Box minWidth="10.5rem">
                              <Typography fontWeight="500" align="center">
                                Stats
                              </Typography>

                              <Box className="specs_container">
                                <Box className="specs">
                                  <Typography>HP</Typography>
                                  <Typography>{data?.hp}</Typography>
                                </Box>

                                <Loader
                                  variant="determinate"
                                  value={normalize(data?.hp)}
                                ></Loader>
                              </Box>
                              <Box className="specs_container">
                                <Box className="specs">
                                  <Typography>ARMOR</Typography>
                                  <Typography>{data?.armor}</Typography>
                                </Box>
                                <Loader
                                  variant="determinate"
                                  value={normalize(data?.armor)}
                                ></Loader>
                              </Box>
                              <Box className="specs_container">
                                <Box className="specs">
                                  <Typography>ATTACK DAMAGE</Typography>
                                  <Typography>{data?.attackdamage}</Typography>
                                </Box>
                                <Loader
                                  variant="determinate"
                                  value={normalize(data?.attackdamage)}
                                ></Loader>
                              </Box>
                              <Box className="specs_container">
                                <Box className="specs">
                                  <Typography>ATTACK RANGE</Typography>
                                  <Typography>{data?.attackrange}</Typography>
                                </Box>

                                <Loader
                                  variant="determinate"
                                  value={normalize(data?.attackrange)}
                                ></Loader>
                              </Box>
                              <Box className="specs_container">
                                <Box className="specs">
                                  <Typography>MOVE SPEED</Typography>
                                  <Typography>{data?.movespeed}</Typography>
                                </Box>
                                <Loader
                                  variant="determinate"
                                  value={normalize(data?.movespeed)}
                                ></Loader>
                              </Box>
                            </Box>
                          </CardContent>
                        </Box>
                      </Card>
                    </div>
                  </div>
                );
              })
            )}
          </Box>
        </Fade>
      </Modal>
    </Fragment>
  );
};

export default HeroDetails;
