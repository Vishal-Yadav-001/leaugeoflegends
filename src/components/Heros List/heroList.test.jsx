import { act, fireEvent, render, screen, waitFor, waitForElementToBeRemoved } from "@testing-library/react";
import axios from "axios";
import { HerosContext, HerosProvider } from "../../context/heroscontext";
import HeroDetails from "../HeroDetails/heroDetails.component";
import HerosList from "./herosList.component";
import { getChampion } from "../../utils/api";
import renderer from "react-test-renderer"
import { useContext } from "react";
describe("Heros List", () => {
  test("display list of heros", () => {
    const tree = render(
      <HerosProvider>
        <HerosList herosList={heroDetails} orderBy={orderBy} />
      </HerosProvider>
    );
    expect(screen.getAllByText("Zilean")[0]?.innerHTML).toBe("Zilean");
  });

  test("display detail of hero selected", () => {
    const tree = render(
      <HerosProvider>
        <HerosList herosList={heroDetails} orderBy={orderBy}/>
      </HerosProvider>
    );
    const viewDetailsButton = screen.getByTestId("view player details");
    act(async() => {
      fireEvent.click(viewDetailsButton);
      const heroDetails = render(
        <HerosProvider>
          <HeroDetails selectedHero={"Zilean"} open={true} />
        </HerosProvider>
      );
      const data = await getChampion("Zilean");
      expect(axios.get).toHaveBeenCalled();
      expect(data).toEqual(heroDetails);
    });
  });
  test("snapshot",()=>{
    const heroDetailsList = renderer.create( <HerosProvider>
      <HerosList herosList={heroDetails} orderBy={orderBy}/>
    </HerosProvider>).toJSON();
    expect(heroDetailsList).toMatchSnapshot();
  })

});
describe("add to watchlist",()=>{
  
  test("add to watchlist",()=>{
    render(
      <HerosProvider>
        <HerosList herosList={heroDetails} orderBy={orderBy}/>
      </HerosProvider>
    );
    const addButton = screen.getByTestId("add");
    act(()=>{
      fireEvent.click(addButton);
       waitFor(()=>{
        screen.queryAllByTestId("add").toBe(null);
       })
    })
  });
  /**
   * Query here because getting render error
   */
//   test("remove from watchlist",()=>{
//     const favoriteHero = heroDetails[0];
//     render(
//       <HerosContext.Provider value={{
//  favoriteHero,checkIsHeroPresent,addHero,handleRemoveRecord}}>
//         <HerosList herosList={heroDetails} orderBy={orderBy}/>
//       </HerosContext.Provider>
//     );
//     expect(screen.queryByTestId("add")).toBe(null);
//     const remove = screen.getByTestId("remove");
//       act(()=>{
//         fireEvent.click(remove);
//          waitFor(()=>{
//           screen.queryAllByTestId("remove").toBe(null);
//          })
//       })
//   });
});
const orderBy = "hp";
const checkIsHeroPresent =(val)=>{return true};
const handleRemoveRecord =(val)=>{return null};
const addHero =(val)=>{return null};
const heroDetails = [
  {
    armor: 24,
    armorperlevel: 5,
    attackdamage: 52,
    attackdamageperlevel: 3,
    attackrange: 550,
    attackspeedoffset: null,
    attackspeedperlevel: 2.13,
    big_image_url:
      "https://cdn.pandascore.co/images/lol/champion/big_image/8842e560-c719-4423-86a1-a47b042ce627.jpg",
    crit: 0,
    critperlevel: 0,
    hp: 574,
    hpperlevel: 96,
    hpregen: 5.5,
    hpregenperlevel: 0.5,
    id: 3238,
    image_url:
      "https://cdn.pandascore.co/images/lol/champion/image/0449c26e-f41d-4289-a2bc-32ddb8b3b911.png",
    movespeed: 335,
    mp: 452,
    mpperlevel: 50,
    mpregen: 11.35,
    mpregenperlevel: 0.8,
    name: "Zilean",
    spellblock: 30,
    spellblockperlevel: 1.3,
    videogame_versions: ["12.21.1"],
  },
];
