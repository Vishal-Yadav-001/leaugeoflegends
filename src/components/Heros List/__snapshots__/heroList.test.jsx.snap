// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heros List snapshot 1`] = `
Array [
  <div
    className="MuiContainer-root MuiContainer-maxWidthLg herosList css-1oqqzyl-MuiContainer-root"
  >
    <div
      className="MuiBox-root css-1177il5"
    >
      <div
        className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root hero_list css-bhp9pd-MuiPaper-root-MuiCard-root"
      >
        <div
          className="MuiCardActions-root MuiCardActions-spacing card_actions css-1t6e9jv-MuiCardActions-root"
        >
          <button
            aria-label="Armor 24"
            aria-labelledby={null}
            className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            data-mui-internal-clone-element={true}
            disabled={false}
            onBlur={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseOver={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <img
              alt="armor"
              className="logo armor"
              src="armor.png"
            />
            <span
              className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            aria-label="Damage 52"
            aria-labelledby={null}
            className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            data-mui-internal-clone-element={true}
            disabled={false}
            onBlur={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseOver={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <img
              alt="damage"
              className="logo damage"
              src="bandage.png"
            />
            <span
              className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          className="MuiCardContent-root hero_image css-46bh2p-MuiCardContent-root"
        >
          <img
            alt="Zilean"
            className="hero_image"
            src="https://cdn.pandascore.co/images/lol/champion/image/0449c26e-f41d-4289-a2bc-32ddb8b3b911.png"
          />
        </div>
        <div
          className="MuiCardContent-root hero_name css-46bh2p-MuiCardContent-root"
        >
          <p
            className="MuiTypography-root MuiTypography-body1 css-mvvmou-MuiTypography-root"
          >
            Zilean
          </p>
          <span
            style={
              Object {
                "fontSize": ".8rem",
              }
            }
          >
             
            hp 574
          </span>
        </div>
        <div
          className="MuiCardActions-root MuiCardActions-spacing card_actions css-1t6e9jv-MuiCardActions-root"
        >
          <button
            aria-label="view player details"
            aria-labelledby={null}
            className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium eye_icon css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            data-mui-internal-clone-element={true}
            data-testid="view player details"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseOver={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium icon css-joerfb-MuiSvgIcon-root"
              data-testid="RemoveRedEyeIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
              />
            </svg>
            <span
              className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <span
            className="MuiTypography-root MuiTypography-span css-37jdci-MuiTypography-root"
          >
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium add_to_watch css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
              data-testid="add"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              watch
              <span
                className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </span>
          <button
            aria-label="Hp 574"
            aria-labelledby={null}
            className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            data-mui-internal-clone-element={true}
            disabled={false}
            onBlur={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseOver={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <img
              alt="health"
              className="logo health"
              src="heart.png"
            />
            <span
              className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="footer MuiBox-root css-0"
  >
    <span>
      Made in 2022 by Vishal Yadav
    </span>
  </div>,
]
`;
